src/lib/mina_base/parties_logic.ml

```diff
@@ -0,0 +1,454 @@
+          Bool.true_
+    in
+    let (party, remaining), local_state =
+      let to_pop =
+        match is_start with
+        | `Compute start_data ->
+            Ps.if_ is_start' ~then_:start_data.parties
+              ~else_:local_state.parties
+        | `Yes start_data ->
+            start_data.parties
+        | `No ->
+            local_state.parties
+      in
+      let party, remaining = Ps.pop to_pop in
+      let transaction_commitment =
+        match is_start with
+        | `No ->
+            local_state.transaction_commitment
+        | `Yes start_data | `Compute start_data ->
+            let on_start =
+              h.perform
+                (Transaction_commitment_on_start
+                   { start_party= party
+                   ; protocol_state_predicate=
+                       start_data.protocol_state_predicate
+                   ; other_parties= remaining })
+            in
+            Transaction_commitment.if_ is_start' ~then_:on_start
+              ~else_:local_state.transaction_commitment
+      in
+      let local_state =
+        { local_state with
+          transaction_commitment
+        ; token_id=
+            Token_id.if_ is_start' ~then_:Token_id.default
+              ~else_:local_state.token_id }
+      in
+      ((party, remaining), local_state)
+    in
+    let local_state =
+      { local_state with
+        parties= remaining
+      ; will_succeed=
+          ( match is_start with
+          | `Yes start_data ->
+              start_data.will_succeed
+          | `No ->
+              local_state.will_succeed
+          | `Compute start_data ->
+              Bool.if_ is_start' ~then_:start_data.will_succeed
+                ~else_:local_state.will_succeed ) }
+    in
+    let a, inclusion_proof =
+      h.perform (Get_account (party, local_state.ledger))
+    in
+    h.perform (Check_inclusion (local_state.ledger, a, inclusion_proof)) ;
+    let predicate_satisfied : Bool.t =
+      h.perform (Check_predicate (is_start', party, a, global_state))
+    in
+    let a', update_permitted =
+      h.perform
+        (Check_auth_and_update_account
+           { is_start= is_start'
+           ; global_state
+           ; party
+           ; account= a
+           ; transaction_commitment= local_state.transaction_commitment
+           ; inclusion_proof })
+    in
+    let party_succeeded =
+      Bool.(
+        protocol_state_predicate_satisfied && predicate_satisfied
+        && update_permitted)
+    in
+    (* The first party must succeed. *)
+    Bool.(assert_ ((not is_start') || party_succeeded)) ;
+    let local_state =
+      { local_state with
+        success= Bool.( && ) local_state.success party_succeeded }
+    in
+    let local_delta =
+      (* TODO: This is wasteful as it repeats a computation performed inside
+         the account update. *)
+      Amount.(h.perform (Balance a) - h.perform (Balance a'))
+    in
+    let party_token = h.perform (Party_token_id party) in
```

Comment:
I've just realised while reviewing #9015 that we should check that this token `<> Token_id.invalid`, that being a placeholder value for requesting ownership of a fresh token.